### PR TITLE
Tasks: 4, 5, 6 and 7

### PR DESCRIPTION
**Pull request by Ananth-Desai**
Task 4: Implement Image Caching
Task 5: Implement Placeholder Images for Search Results
Task 6: Show Images on Search Results
Task 7: Implement Photo Viewer Screen

### CODE REVIEW

File: Flickr/Sources/FlickrAppVCs/SearchResultsVC.swift [hunk 1]
```diff
@@ -6,17 +6,183 @@
...
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    func returnImageView(cell: UICollectionViewCell, indexPath: IndexPath) -> [NSLayoutConstraint] {
+        let imageView = UIImageView()
+        imageView.tag = 10
+        imageView.translatesAutoresizingMaskIntoConstraints = false
+        imageView.contentMode = .scaleAspectFill
+        imageView.clipsToBounds = true
+        imageView.sizeToFit()
```
Comment (kunalgupta022) [inline Flickr/Sources/FlickrAppVCs/SearchResultsVC.swift]: Please use configureView to configure these properties. It is much cleaner syntax and sets `translatesAutoresizingMaskIntoConstraints` by default

File: Flickr/Sources/FlickrAppVCs/SearchResultsVC.swift [hunk 2]
```diff
@@ -6,17 +6,183 @@
...
+        })
+        task.resume()
+    }
+
+    func setupCollectionView() -> [NSLayoutConstraint] {
+        let collectionViewLayout = UICollectionViewFlowLayout()
+        collectionViewLayout.minimumInteritemSpacing = minimumInteritemSpacing
+        collectionViewLayout.minimumLineSpacing = minimumLineSpacing
+        collectionViewLayout.sectionInset = sectionInset
+        collectionViewLayout.itemSize = CGSize(width: view.frame.width / widthDivisor, height: cellHeight)
```
Comment (kunalgupta022) [inline Flickr/Sources/FlickrAppVCs/SearchResultsVC.swift]: Please segregate this out into a separate function. This adds too much complication into a single function.

File: Flickr/Sources/FlickrAppVCs/SearchResultsVC.swift [hunk 3]
```diff
@@ -6,17 +6,183 @@
...
+        let collectionView = UICollectionView(frame: view.frame, collectionViewLayout: collectionViewLayout)
+        collectionView.register(UICollectionViewCell.self, forCellWithReuseIdentifier: cellReuseIdentifier)
+        collectionView.delegate = self
+        collectionView.dataSource = self
+        collectionView.translatesAutoresizingMaskIntoConstraints = false
```
Comment (kunalgupta022) [inline Flickr/Sources/FlickrAppVCs/SearchResultsVC.swift]: Please use configureView

File: Flickr/Sources/FlickrAppVCs/SearchResultsVC.swift [hunk 4]
```diff
@@ -6,17 +6,183 @@
...
+            collectionView.rightAnchor.constraint(equalTo: view.rightAnchor),
+            collectionView.topAnchor.constraint(equalTo: view.topAnchor),
+            collectionView.bottomAnchor.constraint(equalTo: view.bottomAnchor)
+        ]
+    }
+
+    func returnSpinner(cell: UICollectionViewCell) -> [NSLayoutConstraint] {
+        let spinner = UIActivityIndicatorView()
+        spinner.startAnimating()
+        spinner.translatesAutoresizingMaskIntoConstraints = false
```
Comment (kunalgupta022) [inline Flickr/Sources/FlickrAppVCs/SearchResultsVC.swift]: Please use configureView

File: Flickr/Sources/FlickrAppVCs/SearchResultsVC.swift [hunk 5]
```diff
@@ -6,17 +6,183 @@
...
+        Nuke.loadImage(with: photos[indexPath.row], into: imageView)
+        cell.addSubview(imageView)
+
+        return [
+            imageView.leadingAnchor.constraint(equalTo: cell.leadingAnchor),
+            imageView.trailingAnchor.constraint(equalTo: cell.trailingAnchor),
+            imageView.topAnchor.constraint(equalTo: cell.topAnchor),
+            imageView.bottomAnchor.constraint(equalTo: cell.bottomAnchor)
+        ]
+    }
```
Comment (kunalgupta022) [inline Flickr/Sources/FlickrAppVCs/SearchResultsVC.swift]: Why are you setting up collectionViewCell  in ViewCootroller? Shouldn't this be part of the CollelctionViewCell code?

File: Flickr/Sources/FlickrAppVCs/SearchResultsVC.swift [hunk 6]
```diff
@@ -6,17 +6,183 @@
...
+
     override func viewDidLoad() {
         super.viewDidLoad()
-        view.backgroundColor = viewBackgroundColor
+        view.backgroundColor = .white
```
Comment (kunalgupta022) [inline Flickr/Sources/FlickrAppVCs/SearchResultsVC.swift]: Please always use R.color to access color assets. THis allows us to easily change the color scheme of the app from a single location.

**Review by kunalgupta022 — CHANGES_REQUESTED** [review]

File: Flickr/Resources/Assets.xcassets/Heart Filled.imageset/Contents.json [hunk 7]
```diff
@@ -0,0 +1,23 @@
+{
+  "images" : [
+    {
+      "filename" : "Vector.png",
```
Comment (kunalgupta022) [inline Flickr/Resources/Assets.xcassets/Heart Filled.imageset/Contents.json]: Please ensure you rename these assets before adding them to the project. Please set the name to the same value as the Asset in Xcode -> `Heart Filled` in this case.

File: Flickr/Resources/Assets.xcassets/Heart Outlined.imageset/Contents.json [hunk 8]
```diff
@@ -0,0 +1,23 @@
+{
+  "images" : [
+    {
+      "filename" : "Unfilled Heart Icon.png",
```
Comment (kunalgupta022) [inline Flickr/Resources/Assets.xcassets/Heart Outlined.imageset/Contents.json]: Same as above. `HeartOutlined`.

File: Flickr/Sources/FlickrAppVCs/PhotoViewerVC.swift [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+class PhotoViewerVC: UIViewController {
+    private var url: URL
+    private var favouriteState: Bool?
+    private var imageTitle: String?
+    private weak var stackView: UIStackView?
+    private weak var imageView: UIImageView?
+    private weak var favoriteEnabledButton: UIButton?
+    private weak var favoriteDisabledButton: UIButton?
+
+    init(url: URL, title _: String, imageTitle: String) {
```
Comment (kunalgupta022) [inline Flickr/Sources/FlickrAppVCs/PhotoViewerVC.swift]: Why is `title` not used in the VC? And if there really is no use for it. Why not remove it as a parameter?

File: Flickr/Sources/FlickrAppVCs/PhotoViewerVC.swift [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        view.backgroundColor = viewBackgroundColor
+        let imageViewConstraints = setupImageView(stackView: stackView)
+        NSLayoutConstraint.activate(imageViewConstraints)
+        let stackViewConstraints = setupStackView()
+        NSLayoutConstraint.activate(stackViewConstraints)
```
Comment (kunalgupta022) [inline Flickr/Sources/FlickrAppVCs/PhotoViewerVC.swift]: Please activate constraints only once. Activating constraints multiple times has a performance impact.

File: Flickr/Sources/FlickrAppVCs/PhotoViewerVC.swift [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+            stackView.axis = .horizontal
+            stackView.alignment = .fill
+            stackView.distribution = .fill
+            stackView.spacing = 20
+        }
+        self.stackView = stackView
+        let labelConstraints = returnLabel(stackView: stackView)
+        NSLayoutConstraint.activate(labelConstraints)
+        let iconConstraints = returnIcon(stackView: stackView)
+        NSLayoutConstraint.activate(iconConstraints)
```
Comment (kunalgupta022) [inline Flickr/Sources/FlickrAppVCs/PhotoViewerVC.swift]: Please activate these constraints in `viewDidLoad` and only once.

File: Flickr/Sources/FlickrAppVCs/PhotoViewerVC.swift [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+        disabledButton.addTarget(self, action: #selector(clickedFavorite), for: .touchUpInside)
+        favoriteDisabledButton = disabledButton
+
+        stackView.addArrangedSubview(enabledButton)
+        stackView.addArrangedSubview(disabledButton)
+
+        return [
+            enabledButton.topAnchor.constraint(equalTo: stackView.topAnchor),
+            disabledButton.topAnchor.constraint(equalTo: stackView.topAnchor)
+        ]
```
Comment (kunalgupta022) [inline Flickr/Sources/FlickrAppVCs/PhotoViewerVC.swift]: Why do you need two buttons for the favorites button? can you not achieve this with just one button?
Also ensure that each function is responsible for setting up only 1 component of the view.

File: Flickr/Sources/FlickrAppVCs/PhotoViewerVC.swift [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+
+    private func returnIcon(stackView: UIStackView) -> [NSLayoutConstraint] {
+        let enabledButton = UIButton()
+        enabledButton.configureView { button in
+            button.imageView?.contentMode = .scaleAspectFit
+        }
+        let filledIcon = filledHeartIcon
+        enabledButton.isHidden = true
+        enabledButton.setImage(filledIcon, for: .normal)
+        enabledButton.addTarget(self, action: #selector(clickedFavorite), for: .touchUpInside)
```
Comment (kunalgupta022) [inline Flickr/Sources/FlickrAppVCs/PhotoViewerVC.swift]: These properties can be set inside `configureView`

File: Flickr/Sources/FlickrAppVCs/PhotoViewerVC.swift [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+            imageView.contentMode = .scaleAspectFit
+            imageView.backgroundColor = imageViewBackgroundColor
+        }
+        self.imageView = imageView
+        Nuke.loadImage(with: url, into: imageView)
+        view.addSubview(imageView)
+
+        return [
+            imageView.topAnchor.constraint(equalTo: view.topAnchor),
+            imageView.heightAnchor.constraint(equalToConstant: view.frame.height / 2),
```
Comment (kunalgupta022) [inline Flickr/Sources/FlickrAppVCs/PhotoViewerVC.swift:115]: `view.frame.height / 2` is this the best constraint to be set to get the closest to the original image dimensions?

File: Flickr/Sources/FlickrAppVCs/SearchResultsVC.swift [hunk 15]
```diff
@@ -6,17 +6,205 @@
...
 // MARK: Constants
 
-private let navigationBarTitleColor = returnColorPalette().navigationBarTitleColor
-private let viewBackgroundColor = returnColorPalette().viewBackgroundColor
+private let baseSearchUrl = "https://www.flickr.com/services/rest"
+private let imageSearchUrl = "https://live.staticflickr.com"
+private let apiKey = "397717930841a3bd19df470ac48fc84f"
+private let apiMethod = "flickr.photos.search"
+private let format = "json"
+private let noJsonCallback = 1
```
Comment (kunalgupta022) [inline Flickr/Sources/FlickrAppVCs/SearchResultsVC.swift]: Please extract these to a GlobalConstants Structure as Done in Meditation App.

**Review by kunalgupta022 — CHANGES_REQUESTED** [review]

File: Flickr/Sources/FlickrAppVCs/PhotoViewerVC.swift (hunk 14 above)
Comment (Ananth-Desai) [reply Flickr/Sources/FlickrAppVCs/PhotoViewerVC.swift:115]: Each image in the results of the search are of different dimension. So, I've set a imageView height, and the image is being displayed in it.

File: Flickr/Sources/FlickrAppVCs/PhotoViewerVC.swift [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        view.backgroundColor = viewBackgroundColor
+        let imageViewConstraints = setupImageView()
+        let stackViewConstraints = setupStackView()
+        let labelConstraints = returnLabel(stackView: stackView)
+        let iconConstraints = returnIcon(stackView: stackView)
```
Comment (kunalgupta022) [inline Flickr/Sources/FlickrAppVCs/PhotoViewerVC.swift:42]: Why is there a need to pass stackview as a parameter here? Can't we access that directly from self in the function?

File: Flickr/Sources/FlickrAppVCs/PhotoViewerVC.swift [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+import Nuke
+import UIKit
+
+class PhotoViewerVC: UIViewController {
+    private var url: URL
+    private var favoriteState: Bool
+    private var imageTitle: String?
+    private weak var stackView: UIStackView?
+    private weak var imageView: UIImageView?
+    private weak var favoriteButton: UIButton?
```
Comment (kunalgupta022) [inline Flickr/Sources/FlickrAppVCs/PhotoViewerVC.swift]: You can forceCast the views inside a viewController

File: Flickr/Sources/FlickrAppVCs/SearchResultsVC.swift [hunk 18]
```diff
@@ -6,17 +6,200 @@
...
+    }
+
+    func returnImageViewConstraints(imageView: UIImageView, cell: UICollectionViewCell) -> [NSLayoutConstraint] {
+        [
+            imageView.leadingAnchor.constraint(equalTo: cell.leadingAnchor),
+            imageView.trailingAnchor.constraint(equalTo: cell.trailingAnchor),
+            imageView.topAnchor.constraint(equalTo: cell.topAnchor),
+            imageView.bottomAnchor.constraint(equalTo: cell.bottomAnchor)
+        ]
+    }
```
Comment (kunalgupta022) [inline Flickr/Sources/FlickrAppVCs/SearchResultsVC.swift]: Why is this setup inside the ViewController? Please create a custom CollectionViewCell to set this up? That is much more scalable when you may want to have multiple elements in the ViewCell.

File: Flickr/Sources/FlickrAppVCs/SearchResultsVC.swift [hunk 19]
```diff
@@ -6,17 +6,200 @@
...
+        let spinner = UIActivityIndicatorView()
+        spinner.configureView { spinner in
+            spinner.startAnimating()
+        }
+        cell.addSubview(spinner)
+        return [
+            spinner.centerXAnchor.constraint(equalTo: cell.centerXAnchor),
+            spinner.centerYAnchor.constraint(equalTo: cell.centerYAnchor)
+        ]
+    }
```
Comment (kunalgupta022) [inline Flickr/Sources/FlickrAppVCs/SearchResultsVC.swift]: PLease move this setup to a custom collection view cell as well.

File: Flickr/Sources/FlickrAppVCs/SearchResultsVC.swift [hunk 20]
```diff
@@ -6,17 +6,200 @@
...
+}
+
+extension SearchResultsVC: UICollectionViewDataSource, UICollectionViewDelegate, UICollectionViewDelegateFlowLayout {
+    func collectionView(_: UICollectionView, numberOfItemsInSection _: Int) -> Int {
+        photos.isEmpty ? 20 : photos.count
```
Comment (kunalgupta022) [inline Flickr/Sources/FlickrAppVCs/SearchResultsVC.swift]: Why return 20 if no results? Should you not prevent the VC from coming to this screen if there are no results?

File: Flickr/Sources/FlickrAppVCs/SearchTabDelegates.swift [hunk 21]
```diff
@@ -10,3 +10,7 @@ import Foundation
 protocol SearchScreenViewControllerDelegate: AnyObject {
     func didTapSearchButton(searchString: String)
 }
+
+protocol SearchResultsViewControllerDelegate: AnyObject {
+    func didSelectImage(url: URL, title: String, imageTitle: String)
```
Comment (kunalgupta022) [inline Flickr/Sources/FlickrAppVCs/SearchTabDelegates.swift]: Please declare these delegates in the same file as ViewCOntrollers and Coordinators.

**Review by kunalgupta022 — CHANGES_REQUESTED** [review]

**Review by Purushothaman22 — APPROVED** [review]

**Pull request merged**